### PR TITLE
Fix `TorchNMS.nms()` early exit logic

### LINKED ISSUE #22013: serious bug in NMS suppression
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

```python
import torch
from ultralytics.utils.nms import TorchNMS
# Non-working example
boxes= torch.tensor([[  28.0000, 1322.0000,  346.2500, 1798.0000],
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([0.6792, 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 2, 3, 4], device='cuda:0')  --> expect tensor([0, 1, 3], device='cuda:0')

```

### Environment

Ultralytics 8.3.196 🚀 Python-3.10.18 torch-2.8.0+cu128 CUDA:0 (Quadro RTX 8000, 48585MiB)
Setup complete ✅ (20 CPUs, 251.3 GB RAM, 2104.4/3297.7 GB disk)

OS                     Linux-6.5.0-35-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.18
Install                pip
Path                   /home/chuong/miniconda3/envs/tracktrackcpp/lib/python3.10/site-packages/ultralytics
RAM                    251.34 GB
Disk                   2104.4/3297.7 GB
CPU                    Intel Xeon Silver 4210R 2.40GHz
CPU count              20
GPU                    Quadro RTX 8000, 48585MiB
GPU count              4
CUDA                   12.8

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.6>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.0.0
polars                 ✅ 1.33.0
ultralytics-thop       ✅ 2.0.17>=2.0.0

### Minimal Reproducible Example

```python
import torch
from ultralytics.utils.nms import TorchNMS
from torchvision.ops import nms as torch_nms

# Non-working example
boxes= torch.tensor([[  28.0000, 1322.0000,  346.2500, 1798.0000],
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([0.6792, 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 2, 3, 4], device='cuda:0')  ---> wrong.

# Working example
keep_indices = torch_nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 1, 3], device='cuda:0')  --> correct

# Working example
boxes= torch.tensor([
        [ 195,  352,  241,  431],
        [ 194,  352,  241,  431],
        [1160,  545, 1260,  690],
        [1160,  545, 1260,  690],], device='cuda:0')
confidences= torch.tensor([ 0.6021, 0.5830, 0.5601, 0.5391], device='cuda:0')

keep_indices = TorchNMS.nms(boxes, confidences, iou_threshold=0.2)
print(keep_indices)
# output: tensor([0, 2], device='cuda:0')  --> correct
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22013 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Fixes a correctness bug and simplifies sorting in NMS, improving detection reliability across Ultralytics models. ✅

### 📊 Key Changes
- Replaced `scores.sort(...)[1]` with `scores.argsort(...)` to get index order directly (cleaner and slightly more efficient).
- Corrected early-exit behavior: when the top box has no overlap with remaining boxes, the loop now continues processing instead of keeping all remaining boxes unchecked.
- Minor cleanup of masking and comments for clarity.

### 🎯 Purpose & Impact
- Improves NMS correctness: prevents mistakenly keeping overlapping boxes when the current best box has zero overlap with the rest, reducing false positives.
- Provides more stable and accurate detection outputs across YOLO models (e.g., YOLO11) with negligible performance impact.
- Code is cleaner and easier to maintain, aiding future optimizations. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline review; an Ultralytics engineer will also assist you soon. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Since this PR addresses a bug in `ultralytics/utils/nms.py`, please also:
- 🧪 Provide a Minimum Reproducible Example (MRE) that demonstrates the previous incorrect behavior and the corrected outcome after your changes. If possible, include a small snippet and expected outputs to validate the NMS early-exit logic across disjoint boxes.
- 🧭 Add or update unit tests to cover the no-overlap case and ensure future regressions are caught. Confirm tests pass locally and in CI.
- 📈 If available, share a brief before/after note on the performance impact for representative scenarios (especially no-overlap iterations), even a small sample is helpful.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🎉

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
ah seems torch.int32 as index does not compatible with torch==1.8.0, reverting back to int64 in https://github.com/ultralytics/ultralytics/pull/22014/commits/7b67ee667f280f1e3107c85337825d5681eddd57

**Comment by UltralyticsAssistant** [comment]
Absolutely thrilled to merge this! Thank you @Y-T-G and @Laughing-q for the sharp, impactful improvements to NMS. As Leonardo da Vinci said, “Simplicity is the ultimate sophistication.” Your shift to `argsort` and the corrected early-exit logic embody that—cleaner code and more reliable detections across models like YOLO11, all with negligible performance impact.

This will help our community trust results even more and paves the way for smoother future optimizations. Huge appreciation for the thoughtful fixes and clarity upgrades.

Merged: [PR #22014 – Improve NMS correctness and simplify sorting](https://github.com/ultralytics/ultralytics/pull/22014) 🚀
